rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -138,15 +133,20 @@ void client_data_drop(void * data)
 
   // See the comment about the "num_in_flight" class variable in the ClientData class for
   // why we need to do this.
-  // Note: This called could lead to UB since the ClientData * could have been deallocated
-  // if the query reply is received after NodeData was destructed. ie even though we do not
-  // erase the ClientData from NodeData's clients_ map when rmw_destroy_client is called,
-  // the clients_ maps would get erased when the NodeData's destructor is invoked.
-  // This is an edge case that should be resolved once we switch to zenoh-cpp and capture
-  // weaK_ptr<ClientData> in this callback instead.
+  std::lock_guard<std::mutex> lock(num_in_flight_mutex);
   auto num_in_flight_it = num_in_flight_map.find(client_data);
-  if (num_in_flight_it != num_in_flight_map.end()) {
-    --num_in_flight_it->second;
+  if (num_in_flight_it == num_in_flight_map.end()) {
+    // This should never happen
+    RMW_ZENOH_LOG_ERROR_NAMED(
+      "rmw_zenoh_cpp",
+      "Unable to find object in num_in_flight_map."
+    );
+    return;
+  }
+
+  --num_in_flight_it->second;
+  if (num_in_flight_it->second == 0) {
```

Comment:
We only need to remove the client from `deleted_clients` if it was deleted in the past. I guess the call to erase is harmless if the client was not deleted before but this makes the operation semantically clear.
```suggestion
  if (num_in_flight_it->second == 0 && deleted_clients.count(client) > 0) {
```